File: Python/66.py
# 36ms
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        if digits[-1] == 9:
            if len(digits) == 1: #[9]
                return [1, 0]
            return self.plusOne(digits[:-1]) + [0] # last element is now 0
        else:
            digits[-1] += 1
        return digits

# 24ms - 98 faster
class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        for index in range (len(digits) - 1, -1, -1):
            if digits[index] != 9:
                digits[index] += 1
                return digits
            else:
                digits[index] = 0

        return [1] + digits

File: CPP/66.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        for (auto itr = digits.rbegin(); itr != digits.rend(); ++itr){
            if (*itr != 9){
                *itr += 1;
                return digits;
            } else {
                *itr = 0;
            }
        }
        // If the digits length will not be changed, the above loop covers all.
        // In this increment, there will be no case where the carry bit becomes 2,
        // but only one.
        // Therefore, if the digits dont get reutrned in the above loop,
        // all we have to do is just insert the most significant bit.
        digits.insert(digits.begin(), 1);
        return digits;
    }
};
